runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MonitoringInfoMetricName.java

```diff
@@ -113,11 +136,15 @@ public boolean equals(Object o) {
   }
 
   @Override
-  public String toString() {
+  public String toString(String delimiter) {
+    if (getNamespace() != null && getName() != null) {
+      return super.toString(delimiter);
+    }
     StringBuilder builder = new StringBuilder();
-    builder.append(this.urn.toString());
-    builder.append(" ");
-    builder.append(this.labels.toString());
+    if (labels.containsKey(PCOLLECTION_LABEL)) {
```

Comment:
just iterate over the keyas and values and add them to the return string, instead of inspecting the labels for specific keys.

Then there is no need to update this code for new labels.

Reply:
I wrote things that way in several places initially, but ran into interesting problems that we'll need to discuss eventually, so I'll describe them below.

Here's what happens in this PR if I naively append the label kv-pairs: https://github.com/ryan-williams/beam/commit/557628627ae7438a6b688514e81205594238ed0f. 

Note that Flink receives a metric with the unseemly name [`step.PTRANSFORM.step.PCOLLECTION.pcoll.beam.metric.element_count.v1`](https://github.com/ryan-williams/beam/commit/557628627ae7438a6b688514e81205594238ed0f#diff-682be67e606ae40ded3e0087df5f9f8bR136) (which goes e.g. to the Flink web UI, which we've taken pains to make more human-readable).

The first `step` there is added by `MetricKey.toString`, which calls this `MetricName.toString` routine, expecting info about the metric's "name" (URN) only.

## MetricKey vs MetricName

The issue, In my view, is that `MonitoringInfoMetricName` is actually more of a `MetricKey` implementation:
- `MetricKey` ≈ `MonitoringInfoMetricName` ({URN, labels})
  - `stepName` ≈ [labels map with just a `PTRANSFORM` entry]
  - `MetricName` ≈ URN
- `MetricResult` ≈ `MonitoringInfo` ({key, metric value})

I ultimately lay things out in this manner in #7823, folding MIMN into MetricKey.

## Special-casing single-label ptransform/pcollection-scoped metrics

A bigger issue is that I think runners and endpoints are going to want to hard-code [the only 2 cases that we will have for the foreseeable future]: `<ptransform>.<namespace>.<name>` and `<pcollection>.<urn>` (`<ptransform>.<urn>` folds into this nicely as well).

They'll do this for backwards-compat reasons, and/or to make things look nice in e.g. web UIs. I've touched code like this in at least the Samza, Flink, and Spark runners (and that `MetricsHttpSink` we were looking at).

I think it's great that the wire format is extensible to many+new labels, but ended up feeling that the SDK should provide the APIs that deal specifically with the shapes all calling code will care about atm.

Sorry that's a lot! I don't care much what we do in this case, but I think we'll have some harder similar choices shortly 😲 lmk what you think.